### PR TITLE
Improve Prometheus instrumentation for REST API, sources, sinks & processors

### PR DESCRIPTION
**Pull request by jimmidyson**
This reworks the previous Prometheus instrumentation to label sources, sinks & processors so they can be timed independently, as well as adding instrumentation to REST API handlers, labelled appropriately.

### CODE REVIEW

**Comment by jimmidyson** [comment]
@mwielgus @huangyuqi Take a look when you get a chance please.

**Comment by huangyuqi** [comment]
Yes, i will.This instrumentation is more better. :+1:

File: metrics/util/metrics/http.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+	case "OPTIONS", "options":
+		return "options"
+	case "NOTIFY", "notify":
+		return "notify"
+	default:
+		return strings.ToLower(m)
+	}
+}
+
+func sanitizeCode(s int) string {
```
Comment (mwielgus) [inline metrics/util/metrics/http.go]: Why do we need this?

Comment (jimmidyson) [inline metrics/util/metrics/http.go]: Labels are text values in Prometheus metrics, this is converting & (normalizing where necessary) to a string in an efficient way I guess. This is lifted straight from Prometheus codebase but let me do a quick benchmark to see if it's worthwhile.

Comment (mwielgus) [inline metrics/util/metrics/http.go]: I guess that the Heapster API traffic will be in most cases smaller than 10qps and this code is a bit of overkill.

Comment (jimmidyson) [inline metrics/util/metrics/http.go]: Fair enough.

File: metrics/util/metrics/http.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+func (f *fancyResponseWriterDelegator) ReadFrom(r io.Reader) (int64, error) {
+	if !f.wroteHeader {
+		f.WriteHeader(http.StatusOK)
+	}
+	n, err := f.ResponseWriter.(io.ReaderFrom).ReadFrom(r)
+	f.written += n
+	return n, err
+}
+
+func sanitizeMethod(m string) string {
```
Comment (mwielgus) [inline metrics/util/metrics/http.go]: Same as below.

Comment (jimmidyson) [inline metrics/util/metrics/http.go]: Removed.

File: metrics/util/metrics/http.go (hunk 1 above)
Comment (jimmidyson) [inline metrics/util/metrics/http.go]: Removed.

**Comment by mwielgus** [comment]
LGTM

**Pull request merged**